### DISCUSSION TITLE
About typeConverter

### DISCUSSION THREAD

**Discussion by zgid123** [nartc/mapper#515] [Ideas]
Hi, I checked the `typeConverter` example, we have `TimestampString` is extended from `Date` class, so when the mapper runs, those code lines will be skipped because the `TimestampString` is `isDateConstructor`. It won't run the function `getMapping` to get the correct `mapping` from `mapper`.

```ts
        let hasSameIdentifier =
            !isPrimitiveConstructor(destinationMemberIdentifier) &&
            !isDateConstructor(destinationMemberIdentifier) &&
            !isPrimitiveConstructor(sourceMemberIdentifier) &&
            !isDateConstructor(sourceMemberIdentifier) &&
            sourceMemberIdentifier === destinationMemberIdentifier;

        if (hasSameIdentifier) {
            // at this point, we have a same identifier that aren't primitive or date
            // we then check if there is a mapping created for this identifier
            hasSameIdentifier = !getMapping(
                mapper,
                sourceMemberIdentifier as MetadataIdentifier,
                destinationMemberIdentifier as MetadataIdentifier,
                true
            );
        }
```

I am working with `Timestamp` of google protobuf (`google-protobuf/google/protobuf/timestamp_pb`), to pass it to the `gRPC`, we need to do `Timestamp.fromDate(new Date()).toObject()`, the object's structure is `{ seconds: number; nanos: number; }`.

I don't wanna config mapping for this structure and wanna pass to `typeConverter` like this

```ts
class GrpcTimestamp {
  nanos: number;
  seconds: number;
}

createMap(mapper, A, B, typeConverter(GrpcTimestamp, Date, (grpcTimestamp) => {
  const timestamp = new Timestamp();
  const { nanos, seconds } = data;

  timestamp.setNanos(nanos);
  timestamp.setSeconds(seconds);

  return timestamp.toDate();
});
```

But the `hasSameIdentifier` above will be true, and the library will throw error. Don't you think we should have a function or an option to make the `hasSameIdentifier` to be false with the class `GrpcTimestamp`? Or do we have a workaround to make it? Because I wanna reuse the logic to convert all `GrpcTimestamp` to `Date`.

I researched the code and couldn't find a way. Because maybe I will have some classes to work like that in the future and wanna do the `typeConverter` like that. Currently, I extend the `GrpcTimestamp` from `Date` to do the trick.

If we already have a solution, please help.

Thanks.

**Comment by zgid123** [comment]
Actually, the trick is not worked 😂

**Comment by zgid123** [comment]
I just realized that I debugged in wrong place. I will update the status of this discussion later. Currently the function `getMapping` will throw error 

```
Error: Mapping is not found for class GrpcTimestamp extends Date {
    nanos;
    seconds;
} and function Date() { [native code] }
```

**Comment by zgid123** [comment]
after debugging, I found out that it was my mistake when using the `typeConverter`

```ts
export function grpcTimestampToDate(data: GrpcTimestamp): Date {
  const timestamp = new Timestamp();
  const { nanos, seconds } = data;

  timestamp.setNanos(nanos);
  timestamp.setSeconds(seconds.low);

  return timestamp.toDate();
}

createMap(
  mapper,
  GrpcVM,
  VM,
  grpcTimestampToDateConverter,
);
```

it must be

```ts
createMap(
  mapper,
  GrpcVM,
  VM,
  grpcTimestampToDateConverter(),
);
```

I have two functions like that, one function I used correctly, but this one not 😂

**Comment by nartc** [comment]
Hey @zgid123, thanks for the discussion and sorry for the late reply. I halted my OSS work while I was on vacation.

I agree that `isDateConstructor()` check is problematic there. Hm...I'll think more about it.

Did you solve your problem?

**Reply by zgid123** [reply]
yup, it was just my mistake when using wrapper function `grpcTimestampToDateConverter`, that was the reason the log raised the error. So no problem with the current logic.

Thanks for your rely.
